Janis prepare changes

illusional opened the pull request:
Adds:

- JanisTransformation
- Additional extensions
- Guess datatype by filename

Conversion problem: https://janis.readthedocs.io/en/latest/casestudy/conversionproblem.html

illusional commented on janis_core/toolbox/toolbox.py:
```diff
@@ -74,14 +84,48 @@ def add_type(datatype: Type[DataType]) -> bool:
+        JanisShed.hydrate_from(modules)
+
+        JanisShed._has_been_hydrated = True
+        JanisShed._has_hydrated_datatypes = True
```
missing `has_hydrated_tools = True`

illusional commented on janis_core/toolbox/toolbox.py:
```diff
@@ -201,3 +266,47 @@ def process_cls(cls, seen_modules, seen_classes: set, current_layer: int):
         except Exception as e:
             Logger.warn(f"{repr(e)} for type {str(cls)}")
+
+    @staticmethod
```
Remove this method

junyk commented on janis_core/translations/wdl.py:
```diff
@@ -2043,23 +2039,29 @@ def prepare_move_statements_for_input(ti: ToolInput):
+                # move into the current directory
+                dest = "."
+            else:
+                sectag = get_secondary_tag_from_original_tag(ti.id(), s)
```
this `else` could be redundant since `sectag` has been defined with the same definition before `if`

junyk commented on janis_core/workflow/workflow.py:
```diff
@@ -1230,9 +1221,40 @@ def get_dot_plot(
+    def apply_input_documentation(
+        self,
+        inputs: Dict[str, Union[InputDocumentation, str, Dict[str, any]]],
+        should_override=False,
+        strict=False,
```
I think it will be useful to have docstring to briefly explain what `should_override` and `strict` options are for

illusional commented:
Thanks for the review @junyk, I've addressed the feedback from all the review comments. I couldn't work out how to reply to your comments, so I've resolved them :)

junyk commented:
Thanks @illusional  :)
